### DISCUSSION TITLE
15. HTTP 버전 별 특징은 무엇인가요?

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#19] [General]
HTTP 버전 별 특징은 무엇인가요?

**Comment by jongmee** [comment]
- **HTTP/0.9**: 가장 초기 버전으로, GET 메서드만 존재했고 헤더가 없었다. 즉 `HTML 파일만` 전송할 수 있었다.
- **HTTP/1.0**: 요청과 응답에서 `헤더`가 도입되어 Content-Type과 같은 메타데이터를 전송할 수 있고 프로토콜이 `유연성, 확장성`을 가지게 됐다. `상태코드`로 브라우저가 요청의 성공과 실패를 알고 그 결과에 대한 동작을 할 수 있다.
- **HTTP/1.1**: `연결을 재사용`할 수 있어 시간이 절약되고 `파이프라이닝`을 추가해 첫 번째 요청의 응답이 오기 전에 두 번째 요청을 보내 지연 시간이 단축되었다. `Host` 헤더 덕에 같은 IP에서 여러 도메인을 호스트할 수 있고 청크된 응답, 캐시 제어 매커니즘이 도입되었다.
- **HTTP/2**: 텍스트가 아닌 `이진 프로토콜`로 최적화했다. 동일한 연결에서 `병렬 요청`을 수행할 수 있고(= 다중화 프로토콜) `헤더를 압축`한다. 서버가 서버 푸시라는 매커니즘으로 클라이언트 캐시에 데이터를 저장할 수 있다.
- **HTTP/3**: 전송 계층에서 TCP 대신 `QUIC`을 사용해서 더 빠르다. QUIC은 `UDP를 기반으로 한 다중화 프로토콜`로 여러 스트림을 실행하고 각 스트림에 대해 독립적으로 패킷 손실 감지 및 재전송을 구현하므로, 오류가 발생하면 해당 패킷에 데이터가 있는 스트림만 차단한다. (TCP 연결 하나만 사용하는 HTTP/2는 패킷 손실 감지 및 재전송이 모든 스트림을 차단할 수 있다.)


출처: https://developer.mozilla.org/ko/docs/Web/HTTP/Evolution_of_HTTP

**Comment by cutehumanS2** [comment]
1. **HTTP/0.9**
- 사용 가능한 HTTP 메서드가 GET뿐입니다.
- 요청 메시지가 한 줄로만 구성되어 있습니다.
- 헤더가 지원되지 않기 때문에, HTML 파일만 전송될 수 있으며, 다른 유형의 문서는 전송될 수 없음을 의미합니다.

2. **HTTP/1.0**
- HEAD, POST와 같은 GET 이외의 HTTP 메서드가 도입되었습니다.
- 요청과 응답에 **HTTP 헤더가 도입**되어 메타 데이터 전송이 가능해졌고, 프로토콜의 유연성과 확장성이 높아졌습니다.
  - 예를 들어, Content-Type 헤더 덕분에 일반 HTML 파일들 외에 다른 문서들도 전송할 수 있게 되었습니다.
- 각 응답 안에 **상태 코드 라인이 추가**되었습니다.
  - 이를 통해 브라우저가 요청에 대한 성공과 실패를 알 수 있고, 그 결과에 대한 동작을 할 수 있게 되었습니다.
3. **HTTP/1.1**
- **지속 연결**(**keep-alive**)이 공식적으로 지원되었습니다.
  - 하나의 TCP 연결을 재사용할 수 있기 때문에, 여러 HTTP 요청과 응답을 더 빠르게 처리할 수 있습니다.
- **파이프라이닝** 기능이 도입되어, 통신 지연 시간이 단축되었습니다.
  - 특정 요청에 대한 응답이 수신되기 전에 다음 요청을 보낼 수 있습니다.
- **콘텐츠 협상** 기능이 도입되어, 클라이언트와 서버로 하여금 최적의 콘텐츠를 합의하여 제공할 수 있습니다.
- **Host 헤더** 덕분에, 동일 IP 주소에 여러 도메인을 호스팅할 수 있습니다.
  - IP 자원 절약이 가능합니다.
- **청크 전송**이 가능하여, 서버가 응답 본문의 크기를 미리 알 수 없을 때 청크 단위로 응답을 보낼 수 있습니다.
- **Cache-Control 헤더**를 통해 **세밀한 캐싱 정책**을 설정할 수 있습니다.
4. **HTTP/2.0**
- **헤더 압축**(**HPACK**)을 사용해 중복된 헤더 값을 압축하여 전송 효율성과 속도를 높입니다.
- 텍스트가 아닌, **바이너리 데이터 기반의 메시지**를 송수신합니다.
- **서버 푸시** 기능을 제공하여, 클라이언트가 요청하지 않았더라도 미래에 필요할 것으로 예상되는 자원을 미리 전송할 수 있습니다.
- 멀티플렉싱(다중화) 기법을 도입하여, HOL 블로킹 문제를 완화합니다.
  - 멀티플렉싱은 요청별 스트림을 분리하여 병렬적으로 메시지를 주고받는 기술입니다.
5. **HTTP/3.0**
- UDP을 기반으로 구현된 QUIC 프로토콜을 기반으로 동작합니다.
  - QUIC 프로토콜은 UDP 위에 보안(TLS), 연결 유지, 멀티플렉싱 기능을 통합한 전송 프로토콜입니다.
- QUIC는 초기 연결 시 TLS 핸드셰이크를 QUIC 핸드셰이크에 통합하여 3-Way Handshake 없이 빠르게 연결을 수립할 수 있습니다.
- QUIC는 UDP 기반의 멀티플렉싱을 사용하여, 손실된 스트림만 재전송하기 때문에 TCP 기반의 HOL(Head of Line) 블로킹 문제를 근본적으로 해결합니다.

<br>

[참고한 자료]
- <혼자 공부하는 네트워크> - 05-2. HTTP 중 “HTTP의 발전: HTTP/0.9에서 HTTP/3.0까지”
- [MDN - HTTP의 진화](https://developer.mozilla.org/ko/docs/Web/HTTP/Evolution_of_HTTP)
- [MDN - QUIC](https://developer.mozilla.org/ko/docs/Glossary/QUIC)
